Issue with virtual address deletion or replacement when the name field contains the route domain (f.e. %50)

### Environment

 * TMOS/Bigip Version: BIG-IP v15.1.5.1 (Build 0.0.14)
 * Terraform Version: 1.7.2
 * Terraform bigip provider Version: 1.20.2 & 1.16.0

### Summary

We need to append designated route domains to virtual address names as can be seen in the following manifest snippet:

```
resource "bigip_ltm_virtual_address" "virtual_address_buc" {
  for_each = { for vs in var.virtual_servers : vs.name_prefix => vs if contains(var.enable_objects_in, "buc") }

  name            = "/${var.partition}/${each.value.destination}%${each.value.rd}"
  advertize_route = "selective"

  provider = bigip.bigip_buc
}
```
When you populate the name field with values it would look something like this:

```
resource "bigip_ltm_virtual_address" "virtual_address_buc" {

  name            = "/partition/127.0.0.1%50"
  advertize_route = "selective"

  provider = bigip.bigip_buc
}
```

When we have to change the RD in our values for a certain virtual server that automatically triggers a need for a replacement to the virtual address due to the name change. As you can see in the following terraform plan snippet:

```
[2024-02-13 13:24:15]   # module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc["whatever1.foo.bar.intra"] must be replaced
[2024-02-13 13:24:15] -/+ resource "bigip_ltm_virtual_address" "virtual_address_buc" {
[2024-02-13 13:24:15]       ~ id              = "/k8s-devops-devuat/172.**.**.**%50" -> (known after apply)
[2024-02-13 13:24:15]       ~ name            = "/k8s-devops-devuat/172.**.**.**%50" -> "/k8s-devops-devuat/172.**.**.**%52" # forces replacement
[2024-02-13 13:24:15]         # (7 unchanged attributes hidden)
[2024-02-13 13:24:15]     }
```

The issue is that this replacement fails as can be see in the following snippet from terraform apply:

```
[2024-02-13 13:26:53] module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc["whatever1.foo.bar.intra"]: Destroying... [id=/k8s-devops-devuat/172.**.**.**%50]
Error: 01020036:3: The requested Virtual Address (/k8s-devops-devuat/172.**.**.**P) was not found.
```

And the following terraform debug log:

```
2024-02-13T13:26:57.233+0200 [ERROR] provider.terraform-provider-bigip_v1.20.2: Unable to Delete Virtual Address  (/k8s-devops-devuat/172.**.**.**%50) (01020036:3: The requested Virtual Address (/k8s-devops-devuat/172.**.**.**P) was not found.): timestamp="2024-02-13T13:26:57.233+0200"
2024-02-13T13:26:57.233+0200 [TRACE] provider.terraform-provider-bigip_v1.20.2: Called downstream: tf_resource_type=bigip_ltm_virtual_address tf_provider_addr=provider tf_req_id=e7a5e8ba-46ab-d994-f8ca-fc9f9bbe9744 tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:807 @module=sdk.helper_schema timestamp="2024-02-13T13:26:57.233+0200"
2024-02-13T13:26:57.233+0200 [TRACE] provider.terraform-provider-bigip_v1.20.2: Received downstream response: @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tf5serverlogging/downstream_request.go:37 tf_req_duration_ms=3533 diagnostic_error_count=1 tf_proto_version=5.3 tf_provider_addr=provider tf_rpc=ApplyResourceChange @module=sdk.proto diagnostic_warning_count=0 tf_req_id=e7a5e8ba-46ab-d994-f8ca-fc9f9bbe9744 tf_resource_type=bigip_ltm_virtual_address timestamp="2024-02-13T13:26:57.233+0200"
2024-02-13T13:26:57.233+0200 [ERROR] provider.terraform-provider-bigip_v1.20.2: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 diagnostic_summary="01020036:3: The requested Virtual Address (/k8s-devops-devuat/172.**.**.**P) was not found." tf_req_id=e7a5e8ba-46ab-d994-f8ca-fc9f9bbe9744 diagnostic_detail="" @module=sdk.proto diagnostic_severity=ERROR tf_proto_version=5.3 tf_provider_addr=provider tf_resource_type=bigip_ltm_virtual_address tf_rpc=ApplyResourceChange timestamp="2024-02-13T13:26:57.233+0200"
2024-02-13T13:26:57.233+0200 [TRACE] provider.terraform-provider-bigip_v1.20.2: Served request: @module=sdk.proto tf_req_id=e7a5e8ba-46ab-d994-f8ca-fc9f9bbe9744 @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:831 tf_proto_version=5.3 tf_provider_addr=provider tf_resource_type=bigip_ltm_virtual_address tf_rpc=ApplyResourceChange timestamp="2024-02-13T13:26:57.233+0200"
2024-02-13T13:26:57.233+0200 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/f5networks/bigip" is in the global cache
2024-02-13T13:26:57.233+0200 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc["whatever1.foo.bar.intra"]
2024-02-13T13:26:57.233+0200 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc["whatever1.foo.bar.intra"]
2024-02-13T13:26:57.237+0200 [DEBUG] State storage *remote.State declined to persist a state snapshot
2024-02-13T13:26:57.237+0200 [ERROR] vertex "module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc[\"whatever1.foo.bar.intra\"] (destroy)" error: 01020036:3: The requested Virtual Address (/k8s-devops-devuat/172.**.**.**P) was not found.
2024-02-13T13:26:57.237+0200 [TRACE] vertex "module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc[\"whatever1.foo.bar.intra\"] (destroy)": visit complete, with errors
2024-02-13T13:26:57.238+0200 [TRACE] dag/walk: upstream of "module.k8s-devops-devuat.bigip_ltm_virtual_address.virtual_address_buc[\"whatever1.foo.bar.intra\"]" errored, so skipping
```

Upon further investigating this strange situation I have discovered that 50 is also the ASCII code for uppercase P which lead me to the conclusion that somehow the %50 is getting interpreted into P when passed to the bigip API.

### Steps To Reproduce

Steps to reproduce the behavior:

1. Provide terraform resource config which you are facing trouble along with the output of it.

All the information has been provided above

2. To get to know more about the issue, provide terraform debug logs

All the information has been provided above

### Expected Behavior
The expectation was that the virtual address name should not be interpreted if it contains ASCII hex code (as is conducive due to the nature of route domain notation -> %[route-domain-number]

The replacement/deletion of a virtual address is a normal use case that is required for provisioning/deprovisioning operations.

### Actual Behavior
As explained above the issue is that the virtual address name is interpreted and a string such 172.**.**.**%50 becomes 172.**.**.**P when passed to BIGIP by the provider for checking address existence.

BR,
Raoul

--- a/bigip/resource_bigip_as3.go
+++ b/bigip/resource_bigip_as3.go
@@ -144,7 +144,6 @@ func resourceBigipAs3() *schema.Resource {
 			"tenant_name": {
 				Type:        schema.TypeString,
 				Optional:    true,
-				Deprecated:  "this attribute is no longer in use",
 				Description: "Name of Tenant",
 			},
 			"tenant_filter": {
@@ -195,11 +194,18 @@ func resourceBigipAs3Create(ctx context.Context, d *schema.ResourceData, meta in
 	log.Printf("[DEBUG] perApplication:%+v", perApplication)
 
 	if perApplication && len(tenantList) == 0 {
-		tenant, err := GenerateRandomString(10)
-		log.Printf("[DEBUG] tenant name generated:%+v", tenant)
-		if err != nil {
-			return diag.FromErr(fmt.Errorf("could not generate random tenant name"))
+		log.Printf("[INFO] Creating As3 config perApplication : tenant name :%+v", d.Get("tenant_name").(string))
+		var tenant string
+		if d.Get("tenant_name").(string) != "" {
+			tenant = d.Get("tenant_name").(string)
+		} else {
+			tenant, err = GenerateRandomString(10)
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("could not generate random tenant name"))
+			}
 		}
+		log.Printf("[DEBUG] tenant name :%+v", tenant)
+
 		applicationList := client.GetAppsList(as3Json)
 		err, taskID := client.PostPerAppBigIp(as3Json, tenant)
 		log.Printf("[DEBUG] task Id from deployment :%+v", taskID)
@@ -207,12 +213,14 @@ func resourceBigipAs3Create(ctx context.Context, d *schema.ResourceData, meta in
 			return diag.FromErr(fmt.Errorf("posting as3 config failed for tenants:(%s) with error: %v", tenantFilter, err))
 		}
 		tenantCount = append(tenantCount, tenant)
+		_ = d.Set("tenant_filter", tenant)
+		_ = d.Set("tenant_name", tenant)
 		_ = d.Set("tenant_list", tenant)
 		_ = d.Set("task_id", taskID)
 		_ = d.Set("application_list", applicationList)
 		_ = d.Set("per_app_mode", true)
 	} else {
-		log.Printf("[INFO] Creating As3 config for tenants:%+v", tenantList)
+		log.Printf("[INFO] Creating As3 config traditionally for tenants:%+v", tenantList)
 		tenantCount := strings.Split(tenantList, ",")
 		if tenantFilter != "" {
 			log.Printf("[DEBUG] tenantFilter:%+v", tenantFilter)
@@ -244,6 +252,7 @@ func resourceBigipAs3Create(ctx context.Context, d *schema.ResourceData, meta in
 		log.Printf("[DEBUG] ID for resource :%+v", d.Get("tenant_list").(string))
 		_ = d.Set("task_id", taskID)
 		_ = d.Set("per_app_mode", false)
+		_ = d.Set("tenant_name", tenantList)
 	}
 
 	if !client.Teem {
@@ -353,7 +362,7 @@ func resourceBigipAs3Update(ctx context.Context, d *schema.ResourceData, meta in
 	log.Printf("[DEBUG] perApplication:%+v", perApplication)
 	if d.Get("per_app_mode").(bool) {
 		if perApplication && len(tenantList) == 0 {
-			oldTenantList := d.Get("tenant_list").(string)
+			oldTenantList := d.Id()
 			log.Printf("[INFO] Updating As3 Config for tenant:%s with Per-Application Mode:%v", oldTenantList, perApplication)
 			err, task_id := client.PostPerAppBigIp(as3Json, oldTenantList)
 			log.Printf("[DEBUG] task_id from PostPerAppBigIp:%+v", task_id)
@@ -363,6 +372,7 @@ func resourceBigipAs3Update(ctx context.Context, d *schema.ResourceData, meta in
 			// tenantCount = append(tenantCount, tenant)
 			_ = d.Set("tenant_list", oldTenantList)
 			_ = d.Set("task_id", task_id)
+			_ = d.Set("tenant_filter", oldTenantList)
 		} else {
 			if !perApplication {
 				return diag.FromErr(fmt.Errorf("Per-Application should be true in Big-IP Setting"))
@@ -371,7 +381,7 @@ func resourceBigipAs3Update(ctx context.Context, d *schema.ResourceData, meta in
 			}
 		}
 	} else {
-		log.Printf("[INFO] Updating As3 Config for tenants:%s", tenantList)
+		log.Printf("[INFO] Updating As3 Config Traditionally for tenants:%s", tenantList)
 		oldTenantList := d.Get("tenant_list").(string)
 		tenantFilter := d.Get("tenant_filter").(string)
 		if tenantFilter == "" {
@@ -403,7 +413,7 @@ func resourceBigipAs3Update(ctx context.Context, d *schema.ResourceData, meta in
 		log.Printf("[DEBUG] successfulTenants :%+v", successfulTenants)
 		if err != nil {
 			if successfulTenants == "" {
-				return diag.FromErr(fmt.Errorf("Error updating json  %s: %v", tenantList, err))
+				return diag.FromErr(fmt.Errorf("error updating json  %s: %v", tenantList, err))
 			}
 			_ = d.Set("tenant_list", successfulTenants)
 			if len(successfulTenants) != len(tenantList) {
@@ -411,8 +421,13 @@ func resourceBigipAs3Update(ctx context.Context, d *schema.ResourceData, meta in
 			}
 		}
 		_ = d.Set("task_id", taskID)
+		_ = d.Set("tenant_name", tenantList)
 	}
-	createdTenants = d.Get("tenant_list").(string)
+	if d.Get("tenant_filter").(string) != "" {
+		createdTenants = d.Get("tenant_filter").(string)
+	} else {
+		createdTenants = d.Get("tenant_list").(string)
+	}
 	return resourceBigipAs3Read(ctx, d, meta)
 }
 

--- a/bigip/resource_bigip_as3.go
+++ b/bigip/resource_bigip_as3.go
@@ -144,7 +144,7 @@ func resourceBigipAs3() *schema.Resource {
 			"tenant_name": {
 				Type:        schema.TypeString,
 				Optional:    true,
-				Description: "Name of Tenant",
+				Description: "Name of Tenant. This name is used only in the case of Per-Application Deployment. If it is not provided, then a random name would be generated.",
 			},
 			"tenant_filter": {
 				Type:        schema.TypeString,

--- a/bigip/resource_bigip_ltm_virtual_address.go
+++ b/bigip/resource_bigip_ltm_virtual_address.go
@@ -230,6 +230,6 @@ func modifyNameForRouteDomain(name string) string {
 		name = url.PathEscape(name)
 	}
 
-	fmt.Printf("[INFO] updated name %v", name)
+	log.Printf("[INFO] updated name %v", name)
 	return name
 }

--- a/bigip/resource_bigip_ltm_virtual_address.go
+++ b/bigip/resource_bigip_ltm_virtual_address.go
@@ -10,6 +10,8 @@ import (
 	"context"
 	"fmt"
 	"log"
+	"net/url"
+	"strings"
 
 	bigip "github.com/f5devcentral/go-bigip"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
@@ -175,7 +177,7 @@ func resourceBigipLtmVirtualAddressUpdate(ctx context.Context, d *schema.Resourc
 	client := meta.(*bigip.BigIP)
 
 	name := d.Id()
-
+	name = modifyNameForRouteDomain(name)
 	va := hydrateVirtualAddress(d)
 
 	err := client.ModifyVirtualAddress(name, va)
@@ -211,6 +213,7 @@ func resourceBigipLtmVirtualAddressDelete(ctx context.Context, d *schema.Resourc
 		d.SetId("")
 		return nil
 	}
+	name = modifyNameForRouteDomain(name)
 	err := client.DeleteVirtualAddress(name)
 	if err != nil {
 		log.Printf("[ERROR] Unable to Delete Virtual Address  (%s) (%v)", name, err)
@@ -218,4 +221,15 @@ func resourceBigipLtmVirtualAddressDelete(ctx context.Context, d *schema.Resourc
 	}
 	d.SetId("")
 	return nil
+}
+
+func modifyNameForRouteDomain(name string) string {
+	if idx := strings.LastIndex(name, "/"); idx != -1 {
+		name = name[:idx+1] + url.PathEscape(name[idx+1:])
+	} else {
+		name = url.PathEscape(name)
+	}
+
+	fmt.Printf("[INFO] updated name %v", name)
+	return name
 }

--- a/bigip/resource_bigip_ltm_virtual_address_test.go
+++ b/bigip/resource_bigip_ltm_virtual_address_test.go
@@ -25,6 +25,20 @@ resource "bigip_ltm_virtual_address" "test-va" {
 	traffic_group = "/Common/none"
 }
 `
+var TEST_VA_ROUTE_DOMAIN_CONFIG = `
+resource "bigip_ltm_virtual_address" "test-va" {
+	name            = "/Common/1.1.1.1%50"
+	advertize_route = "selective"
+	icmp_echo = "any"
+  }
+`
+var TEST_VA_ROUTE_DOMAIN_CHANGED_CONFIG = `
+resource "bigip_ltm_virtual_address" "test-va" {
+	name            = "/Common/1.1.1.1%50"
+	advertize_route = "selective"
+	icmp_echo = "selective"
+  }
+`
 var TEST_VA_RESOURCE = fmt.Sprintf(TEST_VA_CONFIG, TEST_VA_NAME)
 var TEST_VA_RESOURCE_NAME_CHANGED = fmt.Sprintf(TEST_VA_CONFIG, TEST_VA_NAME_CHANGED)
 
@@ -48,6 +62,35 @@ func TestAccBigipLtmVA_create(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					testCheckVAExists(TEST_VA_NAME, false),
 					testCheckVAExists(TEST_VA_NAME_CHANGED, true),
+				),
+			},
+		},
+	})
+}
+
+func TestAccBigipLtmVATCIssue936(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAcctPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckVAsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: TEST_VA_ROUTE_DOMAIN_CONFIG,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckVAExists("/Common/1.1.1.1%50", true),
+					resource.TestCheckResourceAttr("bigip_ltm_virtual_address.test-va", "name", "/Common/1.1.1.1%50"),
+					resource.TestCheckResourceAttr("bigip_ltm_virtual_address.test-va", "icmp_echo", "any"),
+				),
+			},
+			{
+				Config:    TEST_VA_ROUTE_DOMAIN_CHANGED_CONFIG,
+				PreConfig: func() { testCheckVAExists(TEST_VA_NAME, true) },
+				Check: resource.ComposeTestCheckFunc(
+					testCheckVAExists("/Common/1.1.1.1%50", true),
+					resource.TestCheckResourceAttr("bigip_ltm_virtual_address.test-va", "name", "/Common/1.1.1.1%50"),
+					resource.TestCheckResourceAttr("bigip_ltm_virtual_address.test-va", "icmp_echo", "selective"),
 				),
 			},
 		},

